### ISSUE TITLE
Improve build times

### ISSUE THREAD

**Issue by joaomoreno** [microsoft/vscode#147376]
Here's the timeline of a [recent build](https://dev.azure.com/monacotools/Monaco/_build/results?buildId=166459&view=logs&j=bd1e231b-cef9-51a2-e461-b67194c0a817), which took in total 1h30:

![image](https://user-images.githubusercontent.com/22350/163174604-384a9bd6-dd4b-4137-a236-51eadbc44aa2.png)

This points to two low hanging fruits which can reduce the cascading effect:

- We could start the `macOSUniversal` as soon as `macOS` is done building, right before it starts running tests
- We could sign in parallel to running tests

I am confident that we can get a build down to 1h if we address these two points. Creating an action item list:

- [ ] Split `macOS` into `macOS` and `macOSTests`
- [ ] Create a long-running `Sign` job which, just like Publish, uses DevOps API to fetch unsigned assets and produce signed assets in parallel

---

<details>

<summary>How to create chart</summary>

```js
const AdmZip = require('adm-zip');
const open = require('open');

function main(zipPath) {
  const zip = new AdmZip(zipPath);
  const entries = zip.getEntries()
    .filter(e => !e.isDirectory)
    .filter(e => !/\//.test(e.entryName) && /\.txt$/i.test(e.entryName))

  const tasks = [];

  for (const entry of entries) {
    const startRx = /^(?<timestamp>\S+) ##\[section\]Starting: (?<task>\w+)\s+\S+ ##\[section\]Starting: Initialize job$/mgi;
    const finishRx = /^\S+ ##\[section\]Finishing: Finalize Job\s+(?<timestamp>\S+) ##\[section\]Finishing: (?<task>\w+)$/mgi;

    const contents = entry.getData().toString('utf-8');
    let match = startRx.exec(contents);

    if (!match) {
      continue;
    }

    const start = Date.parse(match.groups['timestamp'].trim());
    const task = match.groups['task'].trim();


    match = finishRx.exec(contents);

    if (!match || match.groups['task'].trim() !== task) {
      continue;
    }

    const end = Date.parse(match.groups['timestamp'].trim());

    tasks.push({ task, start, end });
  }

  const min = Math.min(...tasks.map(t => t.start));

  for (const task of tasks) {
    task.start -= min;
    task.end -= min;
  }

  tasks.sort((a, b) => a.start - b.start);

  const max = Math.max(...tasks.map(t => t.end));
  const scale = 800 / max;

  const svg = `<svg width="840" height="500" xmlns="http://www.w3.org/2000/svg">
    <pattern id="grid" width="100" height="100" patternUnits="userSpaceOnUse">
      <path d="M 100 0 L 0 0 0 100" fill="none" stroke="gray" stroke-width="1"/>
    </pattern>
    <rect width="100%" height="100%" fill="url(#grid)"/>
    ${tasks.map((t, i) => `
      <rect x="${20 + t.start * scale}" y="${20 + i * 20}" width="${(t.end - t.start) * scale}" height="16" style="fill:rgb(0,0,255)" />
      <text x="${22 + t.start * scale}" y="${32 + i * 20}" font-family="Verdana" font-size="12" fill="white">${t.task}</text>
    `).join('')}
  </svg>`;

  const uri = `data:image/svg+xml,${encodeURIComponent(svg)}`;
  open(uri, { app: { name: open.apps.edge } });
}

main(process.argv[2])
```

Place this in a `main.js` file, and run `npm i adm-zip open` in the same folder. Then, download the logs ZIP from a build and point the script to it:

```
$ node main.js LOGS.ZIP
```

![image](https://user-images.githubusercontent.com/22350/163173870-8fceaef2-5d46-4608-a10c-78a7549ddf06.png)

</details>

**Comment by joaomoreno** [comment]
> Create a long-running `Sign` job which, just like Publish, uses DevOps API to fetch unsigned assets and produce signed assets in parallel

This is actually hard on Windows, since signing a setup is so intricately connected with creating the setup itself.

**Comment by lszomoru** [comment]
Would it make sense to break apart building, testing, and signing for each platform? As in as soon as building is completed, we should start running tests and signing is parallel, and the outcome of the tests would just control whether something is published or not.

**Comment by joaomoreno** [comment]
We've talked about that in the past, and each has attempted that once, so no promises there. I want to focus on the thing that is causing the waterfall for now: macOS.

**Comment by TylerLeonhardt** [comment]
Times like these I wish Azure Pipelines had “stage triggers” that could be dependent on a specific artifact showing up rather than having to hand roll our own solution like the Publish stage.

We _could_ try to be a good citizen of Pipelines and just have 3 distinct changes:

* build stage
* test stage
* sign stage

amd then test and sign could happen in parallel. However, there is that cost of having to re-setup the stage.

**Comment by joaomoreno** [comment]
In a 1-1, @TylerLeonhardt just had a brilliant realisation. Splitting build and test apart is costly because test needs pretty much the same amount of overhead as build. The bulk of build is actually the overhead, as built itself takes like 3 mins out of a total of 22 mins. So extracting test would take a significant effort just to shave off 3 minutes altogether. It was at this point that I saw a 💡 above @TylerLeonhardt's head and he said _well, let's just clone the macOS job into a macOSTest job... the former doesn't test and the latter doesn't publish_. In essence, we now have two parallel jobs which both do exactly the same overhead but after that initial part, one builds and publishes and the other tests. This way we can shave off 19 miuntes, with minimal effort. This is happening in https://github.com/microsoft/vscode/tree/joao/build-parallel-darwin.

**Comment by joaomoreno** [comment]
That's a 53m build right there, talk about low hanging fruit!

![image](https://user-images.githubusercontent.com/22350/163320242-1f2865d8-9b33-44bb-ac30-8b6a7a846572.png)

There are two downsides to this:

- We keep two agents busy doing the same work, for like 20m
- We end up shipping bits which haven't _really_ passed the test suites

But this shows that it's possible, will merge this branch. I suggest next step should be to modify the `macOSUniversal` and `macOSSign` jobs to poll for assets, just like `Publish`.

**Comment by joaomoreno** [comment]
For perspective: the bottleneck in the entire macOS stage is tests. The entire stage took 52m. It takes the same amount of time for the tests to run that it takes to do everything else.

![image](https://user-images.githubusercontent.com/22350/163332395-b06c76d6-55ca-47fb-a0cd-787c98e7a250.png)

**Comment by joaomoreno** [comment]
And here's the gantt chart:

![image](https://user-images.githubusercontent.com/22350/163333225-8b7e9f67-d8fa-41ca-aaa6-a0088f2af6e3.png)

🚀

**Comment by lszomoru** [comment]
Inspired by the work that you have been doing for Mac builds, I have taken a quick look at applying your learning to the Windows build. For the latest Insiders release, the Windows stage took 44 minutes to execute. After splitting the build/sign and test into two separate jobs I was able to get the Windows stage down to 33 minutes (ignore the warnings as there was some exceptions while uploading the test logs). Created [a draft PR](https://github.com/microsoft/vscode/pull/147687) in case you would like to look at these changes and take them.

<img width="294" alt="image" src="https://user-images.githubusercontent.com/3372902/163988988-f626afc0-e506-4950-b00a-ffa81be9478b.png">

**Comment by TylerLeonhardt** [comment]
Using `VSCODE_STEP_ON_IT` is clever. The 10min back is nice. Do we think the 10min is worth the fact that we aren't testing the bits that will actually be released? I was already slightly uneasy with making the change for macOS but saving 30min seemed worth it...10min might be a different story. Not really feeling strongly about either choice.

**Comment by joaomoreno** [comment]
> Using `VSCODE_STEP_ON_IT` is clever

Yeah that's a cool trick! Very nice concise changes.

I'm also with mixed feelings. The trade-off now envolves keeping two additional agents busy so we can shave 10 mins off the Window stage. Note that this probably doesn't translate to 10 mins off the entire build, since Windows seems to end approximately at the same time as macOSTest, so we might only shave 3-4 mins off the entire build.

**Comment by lszomoru** [comment]
> The trade-off now envolves keeping two additional agents busy so we can shave 10 mins off the Window stage. 

The 2 additional agents, are only being used when we are doing a full product build. No agent usage changes for CI builds. Also for Windows we are using our own private agents so we are not cannibalising the public Azure DevOps agent pool.

> Note that this probably doesn't translate to 10 mins off the entire build, since Windows seems to end approximately at the same time as macOSTest, so we might only shave 3-4 mins off the entire build.

That is correct, and brings up my next question. Why does the macOSTest job run for such a long time? Based on the chart we are able to build, test, sign Windows in the same amount of time that it takes us to build and test for MacOS.

**Comment by joaomoreno** [comment]
Windows:

![image](https://user-images.githubusercontent.com/22350/164190591-f011ef6d-4a2d-4a0d-9a8c-8e47eb9689ff.png)


macOS:

![image](https://user-images.githubusercontent.com/22350/164190636-605b2d36-6505-4826-a73f-ef4fb60b43ee.png)

---

They're just slower machines.

**Comment by joaomoreno** [comment]
> Especially that `Run integration tests (Remote)` is quite suspicious.

Yup, they seem to be doing different things here... Window starts with:

> $ node --max_old_space_size=8192 ./node_modules/gulp/bin/gulp.js compile-extension:vscode-api-tests compile-extension:microsoft-authentication compile-extension:github-authentication compile-extension:vscode-test-resolver

While macOS does:

> $ node --max_old_space_size=8192 ./node_modules/gulp/bin/gulp.js compile-extension:vscode-api-tests compile-extension:vscode-test-resolver compile-extension:markdown-language-features compile-extension:typescript-language-features compile-extension:emmet compile-extension:git compile-extension:ipynb compile-extension:microsoft-authentication compile-extension:github-authentication compile-extension-media

VERY suspicious. I bet `test-integration.sh` and `test-integration.bat` just got out of sync at some point. cc @bpasero

**Comment by joaomoreno** [comment]
Actually wtf... the remote integration test suites are completely broken. The reason why Windows runs faster is also because it breaks earlier and still the build step stays green:

![image](https://user-images.githubusercontent.com/22350/164192510-a7a5aa5d-664c-4198-a1eb-41b00775ce40.png)

https://dev.azure.com/monacotools/Monaco/_build/results?buildId=167146&view=logs&j=ca7250c6-0e2d-5a52-28a2-5c6aea023c92&t=3a10f1bd-e0fa-5fd5-e2f6-09648e53e7d4&l=607

What.

**Comment by joaomoreno** [comment]
Even macOS ends in disaster:

![image](https://user-images.githubusercontent.com/22350/164192680-634b8de6-52ae-413d-a1fb-1cf2437c9533.png)

https://dev.azure.com/monacotools/Monaco/_build/results?buildId=167146&view=logs&j=a5d99aad-4929-5289-bf4a-26b78032353d&t=d4649a83-a250-559c-eec3-8ca47a8ced50&l=1639

I don't even.

**Comment by joaomoreno** [comment]
Created https://github.com/microsoft/vscode/issues/147755

**Comment by bpasero** [comment]
I think somewhat new is that log files fail to publish:

https://monacotools.visualstudio.com/DefaultCollection/Monaco/_build/results?buildId=167560&view=logs&j=a5d99aad-4929-5289-bf4a-26b78032353d&t=1c2aa4df-3416-5b7a-ccd4-38dbf6ea0f41&l=137

I wonder if we now publish log files while tests are running? Do we execute tests twice now and log to the same location? That would be quite critical to avoid.

**Comment by joaomoreno** [comment]
> I wonder if we now publish log files while tests are running? Do we execute tests twice now and log to the same location?

No, only macOSTest will run tests and publish logs. macOS won't do neither.

This build ran after the refactorings, and everything went just fine:

https://monacotools.visualstudio.com/Monaco/_build/results?buildId=167146&view=results

**Comment by joaomoreno** [comment]
> Could still be a race condition

What do you mean?

> Upload failure. Microsoft.VisualStudio.Services.WebApi.VssServiceResponseException: chunkId does not match chunk. (Expected: 00241EFE74F53035C2B72F77724A5FFF503816127F8A958D6DCAE449CD087D7F01, Actual: E44BB776588A32DDFEDE1003E0D70F61A53D4E8423EF8FA9F3C4F6555ED3502A01)

From the actual error message, this just seems to me like a DevOps transient issue. We will see it dissapear soon enough.

**Comment by bpasero** [comment]
Ah, this is actually a different exception from the one @lszomoru had shared with me earlier:

https://dev.azure.com/monacotools/Monaco/_build/results?buildId=167139&view=logs&j=672276a2-8d3a-5fab-615d-090c51352f92&t=b23e99c8-16d0-5d10-9d46-5f088c0dd16f&l=98

That indicates that a second instance locks the log file on Windows. So this build was maybe from a time when we had 2 agents running the same tests over the same logs dir.

**Comment by lszomoru** [comment]
Even if the `macOS` and `macOSTest` jobs were both running tests (which they are not), these jobs are being run on two different machines, and the jobs run in complete isolation from each other. It seems to me that the log files are still being locked by a process and that is why they cannot be uploaded.

**Comment by TylerLeonhardt** [comment]
I think we've made as much progress as we can here. I hope to see some perf gains when we finally get macOS ARM agents... but there's only so much we can do. @joaomoreno mentioned that we could potentially work with the Teams team to see how they do things to see if there's any lessons there... but they have a bit more resources to manage infra.
